storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala

```diff
@@ -61,10 +66,58 @@ object MergeableStore {
       }
 
     /**
-     * Combine original keys with result after put. Some keys might have dropped,
-     * fill those using missingfn.
+     * Combine original keys with result after put.
      */
-    FutureOps.liftFutureValues(keySet, mPutResultFut, missingfn)
+    FutureOps.liftFutureValues(keySet, mPutResultFut, getFailures)
+  }
+
+  /**
+   * Collects keyed futures, partitioning out the failures.
+   */
+  private[algebra] def collectWithFailures[K, V](fs:Map[K, Future[V]]): (Future[Seq[(K, V)]], Map[K, Future[Nothing]]) = {
+    if (fs.isEmpty) {
+      (Future.value(Seq.empty[(K, V)]), Map.empty[K, Future[Nothing]])
+    } else {
+      val fsSize = fs.size
+      val results = new AtomicReferenceArray[Either[(K,Future[Nothing]), (K, V)]](fsSize)
```

Comment:
what about `AtomicReferenceArray[(K, Either[Throwable, V])]` here instead. Seems easier to follow